### PR TITLE
Simple Eureka/WebClient load balancer

### PR DESCRIPTION
**Pull request by tbak**

### CODE REVIEW

File: titus-ext/eureka/src/main/java/com/netflix/titus/ext/eureka/common/SingleServiceLoadBalancer.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+import com.netflix.discovery.CacheRefreshedEvent;
+import com.netflix.discovery.EurekaClient;
+import com.netflix.discovery.EurekaEvent;
+import com.netflix.discovery.EurekaEventListener;
+import com.netflix.titus.common.runtime.TitusRuntime;
+import com.netflix.titus.common.util.CollectionsExt;
+import com.netflix.titus.common.util.time.Clock;
+
+/**
+ * Helper class to track a collection of servers registered with Eureka with properties helping to determine their health,
```
Comment (corindwyer) [inline titus-ext/eureka/src/main/java/com/netflix/titus/ext/eureka/common/SingleServiceLoadBalancer.java]: Elaborate on the strategy used to load balance traffic.

File: titus-ext/eureka/src/main/java/com/netflix/titus/ext/eureka/common/EurekaUris.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+import java.util.List;
+
+import com.google.common.base.Preconditions;
+import com.netflix.titus.common.util.CollectionsExt;
+import com.netflix.titus.common.util.StringExt;
+import org.springframework.web.util.UriComponentsBuilder;
+
+public class EurekaUris {
+
+    public static final String EUREKA_SCHEMA = "eureka";
```
Comment (fabiokung) [inline titus-ext/eureka/src/main/java/com/netflix/titus/ext/eureka/common/EurekaUris.java]: Typo? `EUREKA_SCHEME`

File: titus-ext/eureka/src/main/java/com/netflix/titus/ext/eureka/common/EurekaUris.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        Preconditions.checkArgument(isEurekaSchema(eurekaUri), "Not Eureka schema in URI: %s", eurekaUri);
+        Preconditions.checkArgument(StringExt.isNotEmpty(eurekaUri.getHost()), "Service name not defined in URI: %s", eurekaUri);
+        return eurekaUri;
+    }
+
+    public static String getServiceName(URI eurekaUri) {
+        return eurekaUri.getHost();
+    }
+
+    public static boolean isEurekaSchema(URI eurekaUri) {
```
Comment (fabiokung) [inline titus-ext/eureka/src/main/java/com/netflix/titus/ext/eureka/common/EurekaUris.java]: `Scheme`?

File: titus-ext/eureka/src/main/java/com/netflix/titus/ext/eureka/common/EurekaUris.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        return EUREKA_SCHEMA.equalsIgnoreCase(eurekaUri.getScheme());
+    }
+
+    public static boolean isSecure(URI eurekaUri) {
+        List<String> secureValue = UriComponentsBuilder.fromUri(eurekaUri).build().getQueryParams().get(EUREKA_SECURE_PARAM);
+        if (CollectionsExt.isNullOrEmpty(secureValue)) {
+            return false;
+        }
+        String value = StringExt.safeTrim(secureValue.get(0));
+        return value.isEmpty() || "true".equalsIgnoreCase(value);
```
Comment (fabiokung) [inline titus-ext/eureka/src/main/java/com/netflix/titus/ext/eureka/common/EurekaUris.java]: minor nit, feel free to ignore: `Boolean.valueOf(value)` may be a bit better here

File: titus-ext/eureka/src/main/java/com/netflix/titus/ext/eureka/common/SingleServiceLoadBalancer.java (hunk 1 above)
Comment (tbak) [reply titus-ext/eureka/src/main/java/com/netflix/titus/ext/eureka/common/SingleServiceLoadBalancer.java]: Added

File: titus-ext/eureka/src/main/java/com/netflix/titus/ext/eureka/common/EurekaUris.java (hunk 2 above)
Comment (tbak) [reply titus-ext/eureka/src/main/java/com/netflix/titus/ext/eureka/common/EurekaUris.java]: Yes, fixing

File: titus-ext/eureka/src/main/java/com/netflix/titus/ext/eureka/common/EurekaUris.java (hunk 4 above)
Comment (tbak) [reply titus-ext/eureka/src/main/java/com/netflix/titus/ext/eureka/common/EurekaUris.java]: Empty value == true, so we can write `eureka://myservice?secure`. Using `Boolean.valueOf(value)` would change the semantic.

Comment (fabiokung) [reply titus-ext/eureka/src/main/java/com/netflix/titus/ext/eureka/common/EurekaUris.java]: I meant only for the second part of the `||`:

```java
return value.isEmpty() || Boolean.valueOf(value)
```

**Pull request merged**